pkg/gadgets/trace/capabilities/tracer/bpf/capable.bpf.c

```diff
@@ -176,6 +184,8 @@ int BPF_KRETPROBE(ig_trace_cap_x)
 
 	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, &event, sizeof(event));
 
+	bpf_map_delete_elem(&start, &pid_tgid);
```

Comment:
If a ebpf program reads a hash map value via the pointer returned by `bpf_map_lookup_elem` after the entry has been deleted via bpf_map_delete_elem, then it might read garbage, unless we use BPF_F_NO_PREALLOC.

Source:
https://lore.kernel.org/lkml/CAG48ez1-WZH55+Wa2vgwZY_hpZJfnDxMzxGLtuN1hG1z6hKf5Q@mail.gmail.com/T/

In this patch, I do both:
- move bpf_map_delete_elem at the end of the function
- use BPF_F_NO_PREALLOC: on my computer there was about 800 threads sleeping in futex/poll syscalls, so the previous limit of 1024 can easily be reached. I increased it to 1024*1024. This map is only used in process context were it is allowed to allocate, so it is safe to use BPF_F_NO_PREALLOC.

Reply:
> This map is only used in process context were it is allowed to allocate, so it is safe to use BPF_F_NO_PREALLOC.

I'm not sure I follow it. AFAIU it was not safe to use BPF_F_NO_PREALLOC (https://github.com/torvalds/linux/commit/94dacdbd5d2d), is there a condition where it's totally safe to use them?

Reply:
Thanks for the link, I didn't know.

What implementation should we do? Some options:
1. Don't use BPF_F_NO_PREALLOC and increases max_entries to a higher value (1024*1024?) and accept the big memory usage
2. Use BPF_MAP_TYPE_TASK_STORAGE but requires Linux 5.11
3. Avoid the sys_enter tracepoint entirely and get the syscall nr via other means (see [example](https://github.com/torvalds/linux/blob/89f5349e0673322857bd432fa23113af56673739/lib/syscall.c#L20-L29))

I would like to try option 3 at some point but for this PR, I will try option 1 to keep the changes small.

Reply:
I think it's fine to go with 1 for the time being.